// grunt-merge-source-maps/tasks/lib/SourceMapMerger.js
var fs = require("fs");
var SourceMap = require("source-map");

var findOriginal = function(consumers, generatedLine, generatedColumn, shouldIgnoreMissingRanges) {
    var currentLine = generatedLine;
    var currentColumn = generatedColumn;
    var original = null;

    consumers.forEach(function(consumer) {
        if (shouldIgnoreMissingRanges && currentLine === null) {
            // We found a mapping that fails to match a statement in the previous
            // item in the compilation chain. Prevent a fatal exception with this
            // check.
            // When forEach completes, findOriginal will return an object with
            // null'd fields. The caller will then exit gracefully.
            return;
        }
        original = consumer.originalPositionFor({
            line: currentLine,
            column: currentColumn
        });
        currentLine = original.line;
        currentColumn = original.column;
    });
    return original;
};

var createMergedSourceMap = function(maps, shouldIgnoreMissingRanges) {
    var consumers = maps.map(function(map) {
        return new SourceMap.SourceMapConsumer(map);
    }).reverse();

    var generator = new SourceMap.SourceMapGenerator({
        file: consumers[0].file
    });

    consumers[0].eachMapping(function(mapping) {
        var original = findOriginal(consumers, mapping.generatedLine, mapping.generatedColumn, shouldIgnoreMissingRanges);
        // source-map uses nulled fields to indicate that it did not find a match.
        if (original.line === null && shouldIgnoreMissingRanges) {
            return;
        }
        // Else: SourceMapGenerator will throw a nice exception for us when we call
        // `addMapping`.

        generator.addMapping({
            generated: {
                line: mapping.generatedLine,
                column: mapping.generatedColumn
            },
            original: {
                line: original.line,
                column: original.column
            },
            source: original.source,
            name: original.name
        });
    });

    return generator.toString();
};

var createMergedSourceMapFromFiles = function(files, shouldIgnoreMissingRanges) {
    var rawDataSets = files.map(function(map) {
        return fs.readFileSync(map);
    });
    var maps = rawDataSets.map(function(data) {
        return JSON.parse(data);
    });

    return createMergedSourceMap(maps, shouldIgnoreMissingRanges);
};

exports.createMergedSourceMap = createMergedSourceMap;
exports.createMergedSourceMapFromFiles = createMergedSourceMapFromFiles;
